src/array.vala

```diff
@@ -92,28 +91,69 @@ public class Array<T>: Object
         return p;
     }
 
-    public unowned T
-    get_scalar(ssize_t [] index)
+    public void*
+    get_pointer ([CCode (array_length = false)] ssize_t[] index)
     {
-        return (T) ((uint8*) _data.get_data () + _offset_for_index (index));
+        return (uint8*) _data.get_data () + _offset_for_index (index);
     }
 
-    public ArrayIterator<T> iterator()
+    public Value
+    get_value ([CCode (array_length = false)] ssize_t[] index)
     {
-        return new ArrayIterator<T>(this);
+        var _value = Value (scalar_type);
+        if (_value.fits_pointer ()) {
+            Memory.copy (_value.peek_pointer (), get_pointer (index), scalar_size);
+        } else if (scalar_type == Type.BOXED) {
+            _value.set_boxed (get_pointer (index));
+        } else if (scalar_type == typeof (string)) {
+            _value.set_string ((string) get_pointer (index));
```

Comment:
OK. Another perspective. We are dealing with two types  : Fixed-length string (char*/char[]) and NULL-terminated string(string).

- Usually it is either NULL-terminated or fixed length. Each has pros and cons. IF we do both we get the minimal common demoniator: unable to represent `\0`, and unable to grow indefinitely.

- get_pointer by definition, returns the pointer to the beginning of the fixed-length string  -- which wasn't a NULL-terminated string to begin with. (It is a fixed-length string). Thus it shall not make any copy. It shall be cast to char*, not string. 

- set_scalar, get_scalar always have to make copies the reason being there is a type cast between fixed length string and NULL-terminated string.
  During the cast to NULL-terminated string, the fixed length string will be truncated at the first `\0`. During the cast to fixed length, though, it doesn't smell good if we leave the extra bytes of the fixed string untouched; because it means we have undefined memory bytes in the fixed length string.